### PR TITLE
KVDB - Design and Write draft of Initial Architecture

### PR DESCRIPTION
**Pull request by JorgeMarinoDev**
|Related issue| Documentation| Testing issue|Wazuh version|Component|
|---|---|---|---|---|
| #16704 | | | 5.0.0 | Engine |

## Description

Initial Architecture Implementation.

## Iterations DoD

 - [x] Iteration 1
   - [x] Study existing code.
   - [x] Write draft Interface Classes and connect to existing code.
   - [x] Identify KVDBHandler's issues. This was the KEY to start a 2nd iteration. 
   - [x] Propose Abstraction of point 1. (Single DB , Multiple CF)
   - [x] Review design with Engine Team.
     - [x] Propose Design for point 4 and fixing point 3.
     - [x] Obtain approvals from lead team members.

 - [x] Iteration 2
   - [x] Exclude kvdb_bench and testing code from compilation. 
   - [x] Comment out OpBuilder KVDB Helper Functions calls to KVDB Manager.
   - [x] Create isolated module to replace existing KVDBManager
   - [x] Wrap-up architecture adding "User Management" or "Actor figure". Remove locking responsibility from KVDBHandler.
   - [x] Get compiling code from scratch.

## Architecture

image

## Files

./include/kvdb2:

    iKVDBHandler.hpp
    iKVDBHandlerManager.hpp
    iKVDBManager.hpp
    iKVDBScope.hpp
    kvdbExcept.hpp
    kvdbHandlerCollection.hpp
    kvdbManagedHandler.hpp
    kvdbManager.hpp
    kvdbScope.hpp
    kvdbSpace.hpp

./src:

    kvdbHandlerCollection.cpp
    kvdbManager.cpp
    kvdbScope.cpp
    kvdbSpace.cpp
## Testing

Unit testing is part of next iteration #17056

### CODE REVIEW

File: src/engine/source/kvdb2/include/kvdb2/diagram.puml
Comment (cborla) [inline src/engine/source/kvdb2/include/kvdb2/diagram.puml:1]: I am not sure about uploading a document file here.

File: src/engine/source/kvdb2/include/kvdb2/iKVDBHandler.hpp [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#ifndef _IKVDBHANDLER_H
```
Comment (cborla) [inline src/engine/source/kvdb2/include/kvdb2/iKVDBHandler.hpp:1]: I think it should be `_I_KVDB_HANDLER_H`

File: src/engine/source/kvdb2/include/kvdb2/iKVDBManager.hpp [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#ifndef _IKVDBMANAGER_H
```
Comment (cborla) [inline src/engine/source/kvdb2/include/kvdb2/iKVDBManager.hpp:1]: I think it should be `_I_KVDB_MANAGER_H`

File: src/engine/source/kvdb2/include/kvdb2/kvdbExcept.hpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
+#ifndef _KVDB_EXCEPT_HPP
```
Comment (cborla) [inline src/engine/source/kvdb2/include/kvdb2/kvdbExcept.hpp:1]: I think it should be `_KVDB_EXCEPT_H`

File: src/engine/source/kvdb2/include/kvdb2/kvdbHandlerCollection.hpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
+#ifndef _KVDBHANDLERCOLLECTION_H
```
Comment (cborla) [inline src/engine/source/kvdb2/include/kvdb2/kvdbHandlerCollection.hpp:1]: I think it should be _KVDB_HANDLER_COLLECTION_H

File: src/engine/source/kvdb2/include/kvdb2/kvdbManagedHandler.hpp [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#ifndef _KVDBMANAGEDHANDLER_H
```
Comment (cborla) [inline src/engine/source/kvdb2/include/kvdb2/kvdbManagedHandler.hpp:1]: I think it should be _KVDB_MANAGED_HANDLER_H

File: src/engine/source/kvdb2/include/kvdb2/kvdbManager.hpp [hunk 6]
```diff
@@ -0,0 +1,91 @@
+#ifndef _KVDBMANAGER_H
```
Comment (cborla) [inline src/engine/source/kvdb2/include/kvdb2/kvdbManager.hpp:1]: I think it should be `_KVDB_MANAGER_H`

File: src/engine/source/kvdb2/include/kvdb2/kvdbScope.hpp [hunk 7]
```diff
@@ -0,0 +1,32 @@
+#ifndef _KVDBSCOPE_H
```
Comment (cborla) [inline src/engine/source/kvdb2/include/kvdb2/kvdbScope.hpp:1]: I think it should be `_KVDB_SCOPE_H`

File: src/engine/source/kvdb2/include/kvdb2/kvdbSpace.hpp [hunk 8]
```diff
@@ -0,0 +1,36 @@
+#ifndef _KVDBSPACE_H
```
Comment (cborla) [inline src/engine/source/kvdb2/include/kvdb2/kvdbSpace.hpp:1]: I think it should be `_KVDB_SPACE_H`

File: src/engine/source/kvdb2/src/kvdbSpace.cpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    if (status.ok())
+    {
+        return value;
+    }
+
+    return base::Error { fmt::format("Cannot get key '{}'. Error: {}", value, key, status.getState()) };
+}
+
+
+} // namespace kvdbManager
```
Comment (cborla) [inline src/engine/source/kvdb2/src/kvdbSpace.cpp:61]: Add a new line at the end.

**Review by cborla — APPROVED** [review]
LGTM!

File: src/engine/source/kvdb2/include/kvdb2/diagram.puml
Comment (JorgeMarinoDev) [reply src/engine/source/kvdb2/include/kvdb2/diagram.puml:1]: We will remove it on future commit.

File: src/engine/source/kvdb2/include/kvdb2/iKVDBHandler.hpp (hunk 1 above)
Comment (JorgeMarinoDev) [reply src/engine/source/kvdb2/include/kvdb2/iKVDBHandler.hpp:1]: Will be fixed in later documentation step.

File: src/engine/source/kvdb2/include/kvdb2/iKVDBManager.hpp (hunk 2 above)
Comment (JorgeMarinoDev) [reply src/engine/source/kvdb2/include/kvdb2/iKVDBManager.hpp:1]: Will be fixed in later documentation step.

File: src/engine/source/kvdb2/include/kvdb2/kvdbExcept.hpp (hunk 3 above)
Comment (JorgeMarinoDev) [reply src/engine/source/kvdb2/include/kvdb2/kvdbExcept.hpp:1]: Will be fixed in later documentation step.

File: src/engine/source/kvdb2/include/kvdb2/kvdbHandlerCollection.hpp (hunk 4 above)
Comment (JorgeMarinoDev) [reply src/engine/source/kvdb2/include/kvdb2/kvdbHandlerCollection.hpp:1]: Will be fixed in later documentation step.

File: src/engine/source/kvdb2/include/kvdb2/kvdbManagedHandler.hpp (hunk 5 above)
Comment (JorgeMarinoDev) [reply src/engine/source/kvdb2/include/kvdb2/kvdbManagedHandler.hpp:1]: Will be fixed in later documentation step.

File: src/engine/source/kvdb2/include/kvdb2/kvdbManager.hpp (hunk 6 above)
Comment (JorgeMarinoDev) [reply src/engine/source/kvdb2/include/kvdb2/kvdbManager.hpp:1]: Will be fixed in later documentation step.

File: src/engine/source/kvdb2/include/kvdb2/kvdbScope.hpp (hunk 7 above)
Comment (JorgeMarinoDev) [reply src/engine/source/kvdb2/include/kvdb2/kvdbScope.hpp:1]: Will be fixed in later documentation step.

File: src/engine/source/kvdb2/include/kvdb2/kvdbSpace.hpp (hunk 8 above)
Comment (JorgeMarinoDev) [reply src/engine/source/kvdb2/include/kvdb2/kvdbSpace.hpp:1]: Will be fixed in later documentation step.

File: src/engine/source/kvdb2/src/kvdbSpace.cpp (hunk 9 above)
Comment (JorgeMarinoDev) [reply src/engine/source/kvdb2/src/kvdbSpace.cpp:61]: Will be fixed in later documentation and cleanup step.

**Pull request merged**